### PR TITLE
Support expanding ~ in $env:PATH when doing command discovery

### LINKED ISSUE #11531: Handle ~ in PATH
# Steps to reproduce

1. Put any executable (ex. `myExe`) at the following path: `~/foo/`
1. Add that path to your PATH
```powershell
$env:PATH = "~/foo$([System.IO.Path]::PathSeparator)$env:PATH"
```
Try to tab complete your executable or just run `myExe`

# Expected behavior

The executable `myExe` is run.

This works just fine in bash and zsh and things like .NET global tools rely on the `~` support in the PATH so we should really fix this.

## caveat

Bash seems to do more - it can handle variables inside of the PATH but that seems like overkill and I can't think of any example that uses that "feature".

We handle tilda pretty much everywhere else, we should be able to support it in application resolution.

# Actual behavior

```none
The term 'myExe' is not recognized as the name of a cmdlet, function, script file, or operable program. Check the spelling of the name, or if a path was included, verify that the path is correct and try again.
```

# Environment data

```none
Name                           Value
----                           -----
PSVersion                      7.0.0-daily.20200108
PSEdition                      Core
GitCommitId                    7.0.0-daily.20200108
OS                             Darwin 18.7.0 Darwin Kernel Version 18.7.0: Sun Dec  1 18:59:03 PST 2019; root:xnu-4903.278.19~1/RELEASE_X86_64
Platform                       Unix
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

### PR DESCRIPTION
**Pull request by SteveL-MSFT**
# PR Summary

The `~` meaning user home path is a commonly used convention.  Bash (and compatible shells) expands this automatically (along with other derivatives not covered in this PR).  Dotnet global tools sets $env:PATH with a path to the tools starting with `~`.  In PowerShell, that is treated as a literal so the tools aren't found.  Fix is in command discovery, when processing $env:PATH, handle the case where a sub-path starts with `~/` or just contains `~` and expand it to be the user home path.

## PR Context

Fix https://github.com/PowerShell/PowerShell/issues/11531

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and enumerated concerns in the summary. This may include:
        - Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode (which runs in a different PS Host).
        - Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
        - Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
        - Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).

### CODE REVIEW

File: src/System.Management.Automation/engine/CommandDiscovery.cs [hunk 1]
```diff
@@ -1328,6 +1328,15 @@ internal LookupPathCollection GetLookupDirectoryPaths()
                     foreach (string directory in tokenizedPath)
                     {
                         string tempDir = directory.TrimStart();
+                        if (tempDir.EqualsOrdinalIgnoreCase("~"))
```
Comment (TylerLeonhardt) [inline src/System.Management.Automation/engine/CommandDiscovery.cs]: This should only check the very first directory, right? I don't know if tilda can be in the middle...

**Review by daxian-dbw — APPROVED** [review]
LGTM

**Comment by daxian-dbw** [comment]
If we do this for `PATH`, shall we do the same for `PSModulePath` then?

**Comment by vexx32** [comment]
There are potentially other PATH values where we might want to apply this as well. I don't recall the exact values off the top of my head, but don't Mac and Linux have PATH-type environment variables that dictate where to look for native library dependencies to load?

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Environment-Variables.Tests.ps1 [hunk 2]
```diff
@@ -42,5 +42,44 @@ Describe "Environment-Variables" -Tags "CI" {
...
-    }
+	}
+
+	Context "~ in PATH" {
+        AfterEach {
+            $env:PATH = $oldPath
+        }
+
+		BeforeAll {
+            $oldPath = $env:PATH
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/Environment-Variables.Tests.ps1]: I guess it is tabs here and below. Please fix indentations.

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Environment-Variables.Tests.ps1 [hunk 3]
```diff
@@ -42,5 +42,44 @@ Describe "Environment-Variables" -Tags "CI" {
...
+			if ($IsWindows) {
+				$pwsh2 = "pwsh2.exe"
+			}
+			else {
+				$pwsh2 = "pwsh2"
+			}
+
+            Copy-Item -Path $pwsh -Destination "~/$pwsh2"
+            $testPath = Join-Path -Path "~" -ChildPath (New-Guid)
+            New-Item -Path $testPath -ItemType Directory
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/Environment-Variables.Tests.ps1]:
```suggestion
            New-Item -Path $testPath -ItemType Directory > $null
```

File: src/System.Management.Automation/engine/CommandDiscovery.cs (hunk 1 above)
Comment (iSazonov) [reply src/System.Management.Automation/engine/CommandDiscovery.cs]: I'd expect that we expand tilde for all elements. There are other applications which add their paths.

**Comment by iSazonov** [comment]
From https://github.com/dotnet/cli/issues/9321 I see that only $HOME works for zsh and MacOs.

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Environment-Variables.Tests.ps1 (hunk 2 above)
Comment (SteveL-MSFT) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/Environment-Variables.Tests.ps1]: I noticed the file was mixed formatting, I'll reformat the entire file.

File: src/System.Management.Automation/engine/CommandDiscovery.cs (hunk 1 above)
Comment (SteveL-MSFT) [reply src/System.Management.Automation/engine/CommandDiscovery.cs]: My intent here is to only handle the beginning case which is the most common case.  `~` in the middle doesn't make sense.

**Comment by SteveL-MSFT** [comment]
@vexx32, you're thinking of `LD_LIBRARY_PATH`.  Similar for `PSModulePath`, I'm not sure if we want to promote usage of this.  This change was specifically to address the dotnet global tool issue.

**Comment by vexx32** [comment]
@SteveL-MSFT yeah I understand that. My concern is simply that we will now have two separate ways we handle the one "kind" of environment variable. 

It's certainly not desirable to promote this oddity too much, I'll agree, but I wonder whether it's better / simpler to be consistent here and have a single way of handling at least initial path resolution for paths in these variables?

File: src/System.Management.Automation/engine/CommandDiscovery.cs [hunk 4]
```diff
@@ -888,7 +888,7 @@ internal static void AutoloadModulesWithJobSourceAdapters(System.Management.Auto
             It attempts to load modules from a fixed ModulesWithJobSourceAdapters list that currently has only `PSScheduledJob` module that is not PS-Core compatible.
```
Comment (PoshChan) [inline src/System.Management.Automation/engine/CommandDiscovery.cs]: @SteveL-MSFT, your last commit had 1 failures in `PowerShell-CI-windows`
Invoke-WebRequest tests.Denial of service.Charset Parsing
```powershell
Expected the actual value to be greater than 10, but got 7.53153622157889.
at <ScriptBlock>, D:\a\1\s\test\powershell\Modules\Microsoft.PowerShell.Utility\WebCmdlets.Tests.ps1: line 1959
1959:             $pathologicalRatio | Should -BeGreaterThan 10
```

**Comment by iSazonov** [comment]
I think we need to document the feature.

**Comment by SteveL-MSFT** [comment]
@PoshChan please retry windows

**Comment by PoshChan** [comment]
@SteveL-MSFT, successfully started retry of `PowerShell-CI-Windows`

**Review by TylerLeonhardt — APPROVED** [review]
LGTM but maybe open an issue to support Tilda in other PATH-like variables.

**Comment by SteveL-MSFT** [comment]
Created https://github.com/PowerShell/PowerShell/issues/11570

**Pull request merged**